tools/ramdump/ramdump_tool.c

```diff
@@ -80,69 +88,221 @@ static int b_read(int fd, uint8_t *buf, int size)
+	/* Array to store heap index corresponding to memory regions */
+	rd_regionx_idx = malloc((number_of_regions + 2) * sizeof(int));
+	if (!rd_regionx_idx) {
+		return -1;
+	}
+	/* Array to map selected regions to be dumped */
+	rd_regionx_map = malloc((number_of_regions + 2) * sizeof(int));
+	if (!rd_regionx_map) {
 		return -1;
 	}
 
-	printf("[>");
-	fflush(stdout);
+	/* Receive memory address, size & heap index for memory regions from TARGET */
+	for (i = 2; i <= number_of_regions + 1; i++) {
+		ret = b_read(dev_fd, (uint8_t *)&mem_address, 4);
+		if (ret != 4) {
+			printf("Receiving address failed, ret = %d\n", ret);
+			return -1;
+		}
+		rd_regionx_start[i] = mem_address;
 
-	while (ramdump_size) {
-		ret = read(dev_fd, &buf, 1);
-		if (ret != 1) {
-			printf("Receiving ramdump %dTH byte failed, ret = %d\n", count, ret);
-			fclose(bin_fp);
+		ret = b_read(dev_fd, (uint8_t *)&mem_size, 4);
+		if (ret != 4) {
+			printf("Receiving size failed, ret = %d\n", ret);
 			return -1;
 		}
+		rd_regionx_size[i] = mem_size;
 
-		ret = fwrite(&buf, 1, 1, bin_fp);
+		ret = read(dev_fd, &c, 1);
 		if (ret != 1) {
-			printf("Writing ramdump %dTH byte failed, ret = %d\n", count, ret);
-			fclose(bin_fp);
+			printf("Receiving size failed, ret = %d\n", ret);
 			return -1;
 		}
+		rd_regionx_idx[i] = c;
+	}
 
-		count++;
-		ramdump_size--;
+	rd_regionx_start[1] = rd_regionx_start[2];
+	rd_regionx_size[1] = rd_regionx_size[2];
 
-		if ((count % (16 * 1024)) == 0) {
-			// printf("%s: received %d bytes\n", __func__, count);
-			printf("\b=>");
-			fflush(stdout);
+	return 0;
+}
+
+static int send_region_info(int dev_fd, char *host_region)
+{
+	int ret;
+	char ack;
+
+	/* Send region info to TARGET */
+	ret = write(dev_fd, host_region, 1);
+	if (ret != 1) {
+		printf("Sending region info failed, ret = %d\n", ret);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int ramdump_recv(int dev_fd)
+{
+	char buf;
+	int i;
+	int ret;
+	int index = 1;
+	int count = 0;
+	int dump_all = 0;
+	int regions_to_dump = 0;
+	uint32_t ramdump_size;
+	char ramdump_region[2] = { '\0' };
+	char bin_file[BINFILE_NAME_SIZE] = { '\0' };
+	FILE *bin_fp;
+
+	printf("\n=========================================================================\n");
+	printf("Ramdump Region Options:\n");
+	printf("1. ALL");
+
+	if (number_of_regions < 2) {
+		printf("\t( Address: %08x, Size: %d )\n",rd_regionx_start[2], rd_regionx_size[2]);
+	} else {
+		for (index = 2; index <= number_of_regions + 1; index++) {
+			printf("\n%d. Region %d:\t( Address: %08x, Size: %d )\t [Heap index = %d]", index, index - 2, rd_regionx_start[index], rd_regionx_size[index], rd_regionx_idx[index]);
+		}
+	}
+	printf("\n=========================================================================\n");
+
+	printf("Please enter desired ramdump option as below:\n \t1 for ALL\n");
+	if (number_of_regions > 1) {
+	printf(" \t2 for Region 0\n \t25 for Region 0 & 3 ...\n");
+	}
+	printf("Please enter your input: ");
+	scanf("%d", &index);
```

Comment:
What happens if user inputs a invalid input or a negative integer.

Reply:
user input needs to be validated before the usage